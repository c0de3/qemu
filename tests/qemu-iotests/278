#!/usr/bin/env bash
#
# Test ref count checks on encrypted images
#
# Copyright (C) 2019 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# creator
owner=berrange@redhat.com

seq=`basename $0`
echo "QA output created by $seq"

status=1	# failure is the default!

_cleanup()
{
	_cleanup_test_img
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./common.rc
. ./common.filter

_supported_fmt qcow2
_supported_proto generic
_supported_os Linux


size=1M

SECRET="secret,id=sec0,data=astrochicken"

IMGSPEC="driver=$IMGFMT,file.filename=$TEST_IMG,encrypt.key-secret=sec0"
QEMU_IO_OPTIONS=$QEMU_IO_OPTIONS_NO_FMT

_run_test()
{
        echo
	echo "== cluster size $csize"
	echo "== checking image refcounts =="
	$QEMU_IMG check --object $SECRET --image-opts $IMGSPEC | _filter_testdir

	echo
	echo "== writing some data =="
	$QEMU_IO --object $SECRET -c "write -P 0x9 0 1" --image-opts $IMGSPEC | _filter_qemu_io | _filter_testdir
	echo
	echo "== rechecking image refcounts =="
	$QEMU_IMG check --object $SECRET --image-opts $IMGSPEC | _filter_testdir

	echo
	echo "== writing some more data =="
	$QEMU_IO --object $SECRET -c "write -P 0x9 $csize 1" --image-opts $IMGSPEC | _filter_qemu_io | _filter_testdir
	echo
	echo "== rechecking image refcounts =="
	$QEMU_IMG check --object $SECRET --image-opts $IMGSPEC | _filter_testdir
}


echo
echo "testing LUKS qcow2 encryption"
echo

for csize in 512 1024 2048 4096 8192 16384 32768 65536
do
  _make_test_img --object $SECRET -o "encrypt.format=luks,encrypt.key-secret=sec0,encrypt.iter-time=10,cluster_size=$csize" $size
  _run_test "driver=$IMGFMT,encrypt.key-secret=sec0,file.filename=$TEST_IMG"
  _cleanup_test_img
done

# success, all done
echo "*** done"
rm -f $seq.full
status=0
