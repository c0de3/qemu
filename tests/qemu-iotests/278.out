QA output created by 278

testing LUKS qcow2 encryption

Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 512
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2078208

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/2048 = 0.05% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2079232

== writing some more data ==
wrote 1/1 bytes at offset 512
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/2048 = 0.10% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2079744
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 1024
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2075648

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/1024 = 0.10% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2077696

== writing some more data ==
wrote 1/1 bytes at offset 1024
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/1024 = 0.20% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2078720
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 2048
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2076672

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/512 = 0.20% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2080768

== writing some more data ==
wrote 1/1 bytes at offset 2048
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/512 = 0.39% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2082816
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 4096
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2084864

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/256 = 0.39% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2093056

== writing some more data ==
wrote 1/1 bytes at offset 4096
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/256 = 0.78% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2097152
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 8192
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2105344

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/128 = 0.78% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2121728

== writing some more data ==
wrote 1/1 bytes at offset 8192
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/128 = 1.56% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2129920
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 16384
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2146304

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/64 = 1.56% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2179072

== writing some more data ==
wrote 1/1 bytes at offset 16384
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/64 = 3.12% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2195456
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 32768
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2228224

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/32 = 3.12% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2293760

== writing some more data ==
wrote 1/1 bytes at offset 32768
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/32 = 6.25% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2326528
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 encrypt.format=luks encrypt.key-secret=sec0 encrypt.iter-time=10

== cluster size 65536
== checking image refcounts ==
No errors were found on the image.
Image end offset: 2359296

== writing some data ==
wrote 1/1 bytes at offset 0
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
1/16 = 6.25% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2490368

== writing some more data ==
wrote 1/1 bytes at offset 65536
1 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

== rechecking image refcounts ==
No errors were found on the image.
2/16 = 12.50% allocated, 0.00% fragmented, 0.00% compressed clusters
Image end offset: 2555904
*** done
